android/sdl_android/src/main/java/com/smartdevicelink/proxy/SdlProxyBase.java

```diff
@@ -2103,7 +2116,102 @@ private void handleErrorsFromInternalMessageDispatcher(String info, Exception e)
 		_proxyListener.onError("Proxy callback dispatcher is down. Proxy instance is invalid.", e);
 	}
 	/************* END Functions used by the Message Dispatching Queues ****************/
-	
+
+
+	private OnRPCNotificationListener mEncryptionReqListener = new OnRPCNotificationListener() {
+		@Override
+		public void onNotified(RPCNotification notification) {
+			List<PermissionItem> permissionItems = ((OnPermissionsChange)notification).getPermissionItem();
+			Set<String> encryptedRpcs = new HashSet<>();
+			Boolean requireEncryptionAppLevel = ((OnPermissionsChange) notification).getRequireEncryption();
+			if (permissionItems != null && !permissionItems.isEmpty()) {
+				for (PermissionItem permissionItem : permissionItems) {
+					if (permissionItem != null) {
+						if (Boolean.TRUE.equals(permissionItem.getRequireEncryption())) {
+							if (requireEncryptionAppLevel == null || requireEncryptionAppLevel == true) {
+								String rpcName = permissionItem.getRpcName();
+								if (rpcName != null) {
+									encryptedRpcs.add(rpcName);
+								}
+							}
+						}
+					}
+				}
+			}
+			if (!encryptedRpcs.isEmpty()) {
+				checkStatusAndInitSecuredService();
+				if (!mEncryptedRPCNames.equals(encryptedRpcs)) {
+					mEncryptedRPCNames = encryptedRpcs;
+				}
+			}
+		}
+	};
+
+	/**
+	 * Checks if an RPC requires encryption
+	 *
+	 * @param rpcName the rpc name (FunctionID) to check
+	 * @return true if the given RPC requires encryption; false, otherwise
+	 */
+	public boolean getRPCRequiresEncryption(@NonNull FunctionID rpcName) {
+		return mEncryptedRPCNames.contains(rpcName.toString());
+	}
+
+	/**
+	 * Gets the encryption requirement
+	 * @return true if encryption is required; false otherwise
+	 */
+	public boolean getRequiresEncryption() {
+		return !mEncryptedRPCNames.isEmpty();
+	}
+
+	/**
+	 * If app is in the foreground and encrypted RPC list is not empty and there is not a secured
+	 * service, start it
+	 */
+	private void checkStatusAndInitSecuredService() {
+		if (!mRPCSecuredServiceStarted && (_hmiLevel != null && _hmiLevel != HMILevel.HMI_NONE) && !mEncryptedRPCNames.isEmpty()) {
+			startProtectedRPCService();
+		}
+	}
+
+	private ISdlServiceListener mRPCSecuredServiceListener = new ISdlServiceListener() {
+		@Override
+		public void onServiceStarted(SdlSession session, SessionType type, boolean isEncrypted) {
+			if(SessionType.RPC.equals(type) && session != null){
+				mRPCSecuredServiceStarted = isEncrypted;

```

Comment:
I am seeing this crash when I test with an app that doesn't set the security library
```
    --------- beginning of crash
2019-09-13 15:22:11.979 19682-19682/io.livio.sdltestsuite.app1 E/AndroidRuntime: FATAL EXCEPTION: main
    Process: io.livio.sdltestsuite.app1, PID: 19682
    java.lang.NullPointerException: Attempt to invoke interface method 'void com.smartdevicelink.managers.ServiceEncryptionListener.onEncryptionServiceUpdated(com.smartdevicelink.protocol.enums.SessionType, boolean, java.lang.String)' on a null object reference
        at com.smartdevicelink.proxy.SdlProxyBase$12.onServiceStarted(SdlProxyBase.java:2183)
        at com.smartdevicelink.SdlConnection.SdlSession.onProtocolSessionStarted(SdlSession.java:604)
        at com.smartdevicelink.protocol.SdlProtocolBase.handleProtocolSessionStarted(SdlProtocolBase.java:1061)
        at com.smartdevicelink.protocol.SdlProtocolBase$MessageFrameAssembler.handleControlFrame(SdlProtocolBase.java:1363)
        at com.smartdevicelink.protocol.SdlProtocolBase$MessageFrameAssembler.handleFrame(SdlProtocolBase.java:1320)
        at com.smartdevicelink.protocol.SdlProtocolBase.handlePacketReceived(SdlProtocolBase.java:667)
        at com.smartdevicelink.protocol.SdlProtocolBase$2.onPacketReceived(SdlProtocolBase.java:1130)
        at com.smartdevicelink.transport.TCPTransportManager$TCPHandler.handleMessage(TCPTransportManager.java:130)
        at android.os.Handler.dispatchMessage(Handler.java:105)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6541)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:240)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:767)
```
We will have to check if the listner is null in `onServiceStarted`, `onServiceEnded ` and `onServiceError `

Reply:
Done